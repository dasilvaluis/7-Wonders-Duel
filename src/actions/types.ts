export const ADD_MONEY = 'ADD_MONEY';
export const TAKE_MONEY = 'TAKE_MONEY';
export const SET_MONEY = 'SET_MONEY';
export const SET_MILITARY_POINTS = 'SET_MILITARY_POINTS';
export const SET_CARDS = 'SET_CARDS';
export const SET_CARD_POSITION = 'SET_CARD_POSITION';
export const SET_WONDERS = 'SET_WONDERS';
export const SET_WONDER_POSITION = 'SET_WONDER_POSITION';
