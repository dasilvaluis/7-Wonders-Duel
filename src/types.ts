export type ElementsMap = {
  [key: string]: GameElement
}

export type Age = 'I' | 'II' | 'III' | 'G';

export enum ElementTypes {
  WONDER_CARD = 'wonder',
  BUILDING_CARD = 'building',
  MILITARY_TOKEN = 'military_token',
  PROGRESS_TOKEN = 'progress_token',
  CONFLICT_PAWN = 'conflict_pawn',
  COIN = 'coin'
}

export interface Position {
  x: number;
  y: number;
}

export interface GameElement extends Position {
  id: string;
  name: string;
  type: ElementTypes;
  faceDown: boolean;
  imageFile: string;
  imageFileBackface: string;
}
