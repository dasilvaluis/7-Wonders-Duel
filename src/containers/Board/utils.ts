import { CARD_MARGIN, BOARD_WIDTH, BOARD_HEIGHT } from '../../contants';
import { Position, GameElement } from '../../types';

export const getRowOf = (howMany: number, cardWidth: number): Array<Position> => {
  const positions = []

  for (let index = 0; index < howMany; index++) {
    positions.push({
      x: (cardWidth + CARD_MARGIN) * index,
      y: 0
    });
  }

  return positions;
};

export const centerRow = (row: Array<Position>, cardsQuantity: number, cardWidth: number) =>
  row.map((position) => ({
    ...position,
    x: position.x - (cardsQuantity * cardWidth + (cardsQuantity - 1) * CARD_MARGIN) / 2
  }));

export const centerHorizontally = (positions: Array<Position>) =>
  movePositions(positions, {
    x: BOARD_WIDTH / 2,
    y: 0
  });

export const centerVertically = (positions: Array<Position>, elementsHeight: number) =>
  movePositions(positions, {
    x: 0,
    y: (BOARD_HEIGHT - elementsHeight)/ 2
  });

export const movePositions = (positions: Array<Position>, offset: Position) => 
  positions.map((position) => ({
    x: position.x + offset.x,
    y: position.y + offset.y
  }));

export const injectPositions = <T>(elements: Array<T>, positions: Array<Position>) =>
  elements.reduce((cards, card, index) => {
    if (index < positions.length) {
      const { x, y } = { x: positions[index].x, y: positions[index].y };

      return [ ...cards, { ...card, x, y } ];
    }

    return cards;
  }, []);

export const shuffleArray = <T>(array: Array<T>) => [ ...array ].sort(() => Math.random() - 0.5);

export const getRandomElements = <T>(array: Array<T>, howMany: number) => {
  const shuffledElements = shuffleArray<T>(array);
  const limit = howMany < shuffledElements.length ? howMany : shuffledElements.length;

  return shuffledElements.slice(0, limit);
};

export const flipCards = (cards: Array<GameElement>) => cards.map((card) => ({ ...card, faceDown: true }));