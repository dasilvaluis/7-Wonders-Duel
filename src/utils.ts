
import { v4 as uuidv4 } from 'uuid';
import { ELEMENT_MARGIN, BOARD_WIDTH } from './contants';
import { GameElement, ElementTypes, Position } from './types';

export const getRowOf = (howMany: number, cardWidth: number): Array<Position> => {
  const positions = []

  for (let index = 0; index < howMany; index++) {
    positions.push({
      x: (cardWidth + ELEMENT_MARGIN) * index,
      y: 0
    });
  }

  return positions;
};

export const centerRow = (row: Array<Position>, cardsQuantity: number, cardWidth: number) =>
  row.map((position) => ({
    ...position,
    x: position.x - (cardsQuantity * cardWidth + (cardsQuantity - 1) * ELEMENT_MARGIN) / 2
  }));

export const centerHorizontally = (positions: Array<Position>) =>
  movePositions(positions, {
    x: BOARD_WIDTH / 2,
    y: 0
  });

export const movePositions = (positions: Array<Position>, offset: Position): Array<Position> => 
  positions.map((position) => ({
    x: position.x + offset.x,
    y: position.y + offset.y
  }));

export const injectPositions = <T>(elements: Array<T>, positions: Array<Position>) =>
  elements.reduce((cards, card, index) => {
    if (index < positions.length) {
      const { x, y } = { x: positions[index].x, y: positions[index].y };

      return [ ...cards, { ...card, x, y } ];
    }

    return cards;
  }, []);

export const shuffleArray = <T>(array: Array<T>) => [ ...array ].sort(() => Math.random() - 0.5);

export const getRandomElements = <T>(array: Array<T>, howMany?: number) => {
  const shuffledElements = shuffleArray<T>(array);
  const limit = howMany && howMany < shuffledElements.length ? howMany : shuffledElements.length;

  return shuffledElements.slice(0, limit);
};

export const flipCards = (cards: Array<GameElement>) => cards.map((card) => ({ ...card, faceDown: true }));

export const keyBy = <ArrayT>(array: Array<ArrayT>, key: string) => 
  (array || []).reduce((r, x) => ({ ...r, [key ? x[key] : x]: x }), {});

export const flattenDeep = <T>(array): Array<T> => Array.isArray(array)
  ? array.reduce((acc: Array<any>, curr) => acc.concat(flattenDeep(curr)) , [])
  : [ array ]

export const createElement = (type: ElementTypes) => ({
  type,
  id: uuidv4(),
  x: 0,
  y: 0,
  faceDown: false,
  imageFile: '',
  imageFileBackface: ''
});

const getElementRealSize = (elementType: ElementTypes) => {
  switch (elementType) {
    case ElementTypes.WONDER_CARD:
      return {
        width: 100,
        height: 66
      }

    case ElementTypes.BUILDING_CARD:
      return {
        width: 44,
        height: 68
      }
    
    case ElementTypes.MILITARY_TOKEN_5:
      return {
        width: 45,
        height: 20
      }
  
    case ElementTypes.MILITARY_TOKEN_2:
      return {
        width: 38,
        height: 17
      }
    
    case ElementTypes.PROGRESS_TOKEN:
      return {
        width: 35,
        height: 35
      }
    
    case ElementTypes.CONFLICT_PAWN:
      return {
        width: 14,
        height: 45
      }
    case ElementTypes.COIN_1:
      return {
        width: 20,
        height: 20
      }
    
    case ElementTypes.COIN_3:
      return {
        width: 25,
        height: 25
      }
    
    case ElementTypes.COIN_6:
      return {
        width: 30,
        height: 30
      }
    
    case ElementTypes.BOARD:
      return {
        width: 390,
        height: 110
      }
    default:
      return {
        width: 0,
        height: 0
      }
  }
};

export const getElementScale = (elementType: ElementTypes) => {
  switch (elementType) {
    case ElementTypes.BUILDING_CARD:
    case ElementTypes.WONDER_CARD:
      return 2.84;
    case ElementTypes.COIN_1:
    case ElementTypes.COIN_3:
    case ElementTypes.COIN_6:
      return 3.33;
    case ElementTypes.CONFLICT_PAWN:
    case ElementTypes.MILITARY_TOKEN_5:
    case ElementTypes.MILITARY_TOKEN_2:
    case ElementTypes.PROGRESS_TOKEN:
    case ElementTypes.BOARD:
      return 1.9;
    default:
      return 1;
  }
};

export const getElementSize = (elementType: ElementTypes) => {
  const { width, height } = getElementRealSize(elementType);
  const scale = getElementScale(elementType);

  return {
    width: width * scale,
    height: height * scale
  };
};

export const getElementStyles = (elementType: ElementTypes): React.CSSProperties => {
  const { width, height } = getElementSize(elementType);
  
  switch (elementType) {
    case ElementTypes.WONDER_CARD:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }

    case ElementTypes.BUILDING_CARD:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }
    
    case ElementTypes.MILITARY_TOKEN_5:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }

    case ElementTypes.MILITARY_TOKEN_2:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }
    
    case ElementTypes.PROGRESS_TOKEN:
      return {
        width: `${ width }px`,
        height: `${ height }px`,
        borderRadius: '50%'
      }
    
    case ElementTypes.CONFLICT_PAWN:
      return {
        width: `${ width }px`,
        height: `${ height }px`,
        borderRadius: 'initial',
        boxShadow: 'initial'
      }
    case ElementTypes.COIN_1:
      return {
        width: `${ width }px`,
        height: `${ height }px`,
        borderRadius: '50%'
      }
    
    case ElementTypes.COIN_3:
      return {
        width: `${ width }px`,
        height: `${ height }px`,
        borderRadius: '50%'
      }
    
    case ElementTypes.COIN_6:
      return {
        width: `${ width }px`,
        height: `${ height }px`,
        borderRadius: '50%'
      }
    
    case ElementTypes.BOARD:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }
    default:
      return {
        width: `${ width }px`,
        height: `${ height }px`
      }
  }
};
