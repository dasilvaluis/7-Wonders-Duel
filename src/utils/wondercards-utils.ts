import {
  centerHorizontally,
  getRowOfCards,
  centerRow,
  getRandomElements,
  movePositions,
  flipCards,
  injectPositions,
  getElementSize,
  flattenDeep
} from './utils';
import { ELEMENT_MARGIN } from '../contants';
import { Coordinates, GameElement, ElementTypes } from '../types';
import { wonders } from '../data/wonders.json';
import { v4 as uuidv4 } from 'uuid';

const moveRowVertically = (row: Array<Coordinates>, rowIndex: number) =>
  row.map((position) => ({
    ...position,
    y: rowIndex * (getElementSize(ElementTypes.WONDER_CARD).height + ELEMENT_MARGIN)
  }));

export const getWonderCardsPlacement = (cardWidth: number) => {
  const rows = [4, 4].map((howMany, rowIndex) => {
    const row = getRowOfCards(howMany, cardWidth);
    const rowMovedVertically = moveRowVertically(row, rowIndex);
    const rowCentered = centerRow(rowMovedVertically, howMany, cardWidth);

    return rowCentered;
  });

  return centerHorizontally(flattenDeep<Coordinates>(rows));
};

export const getShuffledCards = (): Array<GameElement> => 
  getRandomElements(wonders, 8).map((card) => ({
    id: uuidv4(),
    type: ElementTypes.WONDER_CARD,
    x: 0,
    y: 0,
    faceDown: false,
    imageFile: card.file,
    imageFileBackface: 'wonder-back.jpg'
  }));

export const generateWonderCards = (): Array<GameElement> => {
  const wonderWidth = getElementSize(ElementTypes.WONDER_CARD).width;
  const wonderCards = getShuffledCards();
  const cardsPlacement = getWonderCardsPlacement(wonderWidth);
  const cardsPlacementShifted = movePositions(cardsPlacement, {
    y: ELEMENT_MARGIN + getElementSize(ElementTypes.BOARD).height
  });
  const cards: Array<GameElement> = flipCards(injectPositions(wonderCards, cardsPlacementShifted));

  return cards;
};
