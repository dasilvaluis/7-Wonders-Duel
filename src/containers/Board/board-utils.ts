import { GameElement, ElementTypes, Position } from "../../types";
import { tokens as tokensDb } from '../../data/progress-tokens.json';
import { getRandomElements, createElement, getElementScale, getElementSize, movePositions } from "../../utils";
import { CARD_MARGIN, BOARD_WIDTH } from "../../contants";

const boardPosition = {
  x: (BOARD_WIDTH - getElementSize(ElementTypes.BOARD).width) / 2,
  y: CARD_MARGIN,
};

export const getBoardElement = (): GameElement => ({
  ...boardPosition,
  id: 'board',
  type: ElementTypes.BOARD,
  faceDown: false,
  imageFile: 'board.png',
  imageFileBackface: 'board-back.png'
});

export const getProgressTokens = (): Array<GameElement> => {
  const boardTokensCount = 5;
  const tokenWidth = getElementSize(ElementTypes.PROGRESS_TOKEN).width;
  const tokenScale = getElementScale(ElementTypes.PROGRESS_TOKEN);
  const boardScale = getElementScale(ElementTypes.BOARD);
  const shuffledTokens = getRandomElements(tokensDb);
  const boardTokens = shuffledTokens.slice(0, boardTokensCount);
  const hiddenTokens = shuffledTokens.slice(boardTokensCount);

  const boardTokensPlacement = boardTokens.map((el, index) => ({
    x: (tokenWidth + 5 * tokenScale) * index,
    y: 0
  }));

  const placedBoardTokens: Array<GameElement> = boardTokens.map((token , index) => ({
    ...createElement(ElementTypes.PROGRESS_TOKEN),
    x: boardPosition.x + 95 * boardScale + boardTokensPlacement[index].x,
    y: boardPosition.y + 3.75 * boardScale,
    imageFile: token.file,
    imageFileBackface: 'token-progress-back.jpg'
  }));

  const placedHiddenTokens: Array<GameElement> = hiddenTokens.map((token , index) => ({
    ...createElement(ElementTypes.PROGRESS_TOKEN),
    faceDown: true,
    x: boardPosition.x + tokenWidth,
    y: boardPosition.y + 3 * boardScale,
    imageFile: token.file,
    imageFileBackface: 'token-progress-back.jpg'
  }));

  return [ ...placedBoardTokens, ...placedHiddenTokens ];
};

export const getMilitaryTokens = (): Array<GameElement> => {
  const boardScale = getElementScale(ElementTypes.BOARD);
  const token5Height = getElementSize(ElementTypes.MILITARY_TOKEN_5).height;
  const token2Height = getElementSize(ElementTypes.MILITARY_TOKEN_2).height;

  const getMilitaryToken = (value) => {
    const imageFileBackface = 'token-military-back.jpg';

    switch (value) {
      case 5:
        return {
          ...createElement(ElementTypes.MILITARY_TOKEN_5),
          imageFileBackface,
          imageFile: 'token-military-5.jpg'
        };
      default:
        return {
          ...createElement(ElementTypes.MILITARY_TOKEN_2),
          imageFileBackface,
          imageFile: 'token-military-2.jpg'
        };
    }
  };

  const tokensPlacement: Array<Position> = [
    {
      x: 0,
      y: 0
    },
    {
      x: 60 * boardScale,
      y: (token5Height - token2Height) / 2,
    },
    {
      x: 210 * boardScale,
      y: (token5Height - token2Height) / 2,
    },
    {
      x: 263 * boardScale,
      y: 0,
    }
  ];
  
  const tokensPlacementShifted = movePositions(tokensPlacement, {
    x: boardPosition.x + 40 * boardScale,
    y: boardPosition.y + 81.5 * boardScale
  });

  const tokens = [5, 2, 2, 5].map((el, index) => ({
    ...getMilitaryToken(el),
    ...tokensPlacementShifted[index]
  }));

  return tokens;
};

export const getConflictPawn = (): GameElement => {
  const { width: pawnWidth, height: pawnHeight } = getElementSize(ElementTypes.CONFLICT_PAWN);
  const boardWidth  = getElementSize(ElementTypes.BOARD).width;

 return {
    ...createElement(ElementTypes.CONFLICT_PAWN),
    x: boardPosition.x + (boardWidth - pawnWidth) / 2 - 4,
    y: boardPosition.y + pawnHeight - 8,
    imageFile: 'conflict-pawn.png',
    imageFileBackface: 'conflict-pawn-back.png'
  }
};