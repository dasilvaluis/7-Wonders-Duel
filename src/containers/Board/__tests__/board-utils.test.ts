import { getConflictPawn, getMilitaryTokens, getProgressTokens, getBoardElement } from "../board-utils";
import * as utils from '../../../utils';
import { ElementTypes } from "../../../types";

describe('board-utils', () => {
  beforeAll(() => {
    jest
      .spyOn(utils, 'createElement')
      .mockImplementation((el) => ({ 
        id: '',
        x: 0,
        y: 0,
        type: ElementTypes.MOCK,
        faceDown: false,
        imageFile: '',
        imageFileBackface: ''
      }));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('getConflictPawn', () => {
    it('matches snapshot', () => {
      expect(getConflictPawn()).toMatchSnapshot();
    });
  });

  describe('getMilitaryTokens', () => {
    it('matches snapshot', () => {
      expect(getMilitaryTokens()).toMatchSnapshot();
    });
  });

  describe('getProgressTokens', () => {
    it('matches snapshot', () => {
      jest
        .spyOn(utils, 'getRandomElements')
        .mockImplementation((array, limit) => array.slice(0, limit));

      expect(getProgressTokens()).toMatchSnapshot();
    });
  });

  describe('getBoardElement', () => {
    it('matches snapshot', () => {
      expect(getBoardElement()).toMatchSnapshot();
    });
  });
});