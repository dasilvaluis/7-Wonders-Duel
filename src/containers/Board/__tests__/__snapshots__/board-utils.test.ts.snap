// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`board-utils getBoardElement matches snapshot 1`] = `
Object {
  "faceDown": false,
  "id": "board",
  "imageFile": "board.png",
  "imageFileBackface": "board.png",
  "type": "board",
  "x": 488.75,
  "y": 20,
}
`;

exports[`board-utils getConflictPawn matches snapshot 1`] = `
Object {
  "faceDown": false,
  "id": "",
  "imageFile": "conflict-pawn.png",
  "imageFileBackface": "conflict-pawn-back.png",
  "type": "mock",
  "x": 813.75,
  "y": 90.75,
}
`;

exports[`board-utils getMilitaryTokens matches snapshot 1`] = `
Array [
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-military-5.jpg",
    "imageFileBackface": "token-military-back.jpg",
    "type": "mock",
    "x": 558.75,
    "y": 162.625,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-military-2.jpg",
    "imageFileBackface": "token-military-back.jpg",
    "type": "mock",
    "x": 663.75,
    "y": 165.25,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-military-2.jpg",
    "imageFileBackface": "token-military-back.jpg",
    "type": "mock",
    "x": 926.25,
    "y": 165.25,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-military-5.jpg",
    "imageFileBackface": "token-military-back.jpg",
    "type": "mock",
    "x": 1019,
    "y": 162.625,
  },
]
`;

exports[`board-utils getProgressTokens matches snapshot 1`] = `
Array [
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-progress-agricultura.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 655,
    "y": 26.5625,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-progress-albanileria.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 725,
    "y": 26.5625,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-progress-arquitectura.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 795,
    "y": 26.5625,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-progress-derecho.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 865,
    "y": 26.5625,
  },
  Object {
    "faceDown": false,
    "id": "",
    "imageFile": "token-progress-economia.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 935,
    "y": 26.5625,
  },
  Object {
    "faceDown": true,
    "id": "",
    "imageFile": "token-progress-estrategia.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 550,
    "y": 25.25,
  },
  Object {
    "faceDown": true,
    "id": "",
    "imageFile": "token-progress-filosofia.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 550,
    "y": 25.25,
  },
  Object {
    "faceDown": true,
    "id": "",
    "imageFile": "token-progress-matematicas.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 550,
    "y": 25.25,
  },
  Object {
    "faceDown": true,
    "id": "",
    "imageFile": "token-progress-teologia.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 550,
    "y": 25.25,
  },
  Object {
    "faceDown": true,
    "id": "",
    "imageFile": "token-progress-urbanismo.jpg",
    "imageFileBackface": "token-progress-back.jpg",
    "type": "mock",
    "x": 550,
    "y": 25.25,
  },
]
`;
